actors/miner/src/lib.rs

```diff
@@ -3911,9 +3699,194 @@ fn extend_non_simple_qap_sector(
+                update.replica_proof.len(),
+                update.sector_number,
+            );
+            continue;
+        }
+
+        if update.deals.is_empty() {
+            info!("must have deals to update, skipping sector {}", update.sector_number,);
+            continue;
+        }
+
+        if update.deals.len() as u64 > sector_deals_max(policy, sector_size) {
+            info!("more deals than policy allows, skipping sector {}", update.sector_number,);
+            continue;
+        }
+
+        if update.deadline >= policy.wpost_period_deadlines {
+            info!(
+                "deadline {} not in range 0..{}, skipping sector {}",
+                update.deadline, policy.wpost_period_deadlines, update.sector_number
+            );
+            continue;
+        }
+
+        if !is_sealed_sector(&update.new_sealed_cid) {
+            info!(
+                "new sealed CID had wrong prefix {}, skipping sector {}",
+                update.new_sealed_cid, update.sector_number
+            );
+            continue;
+        }
+
+        // Disallow upgrading sectors in immutable deadlines.
+        if !deadline_is_mutable(
+            policy,
+            state.current_proving_period_start(policy, curr_epoch),
+            update.deadline,
+            curr_epoch,
+        ) {
+            info!(
+                "cannot upgrade sectors in immutable deadline {}, skipping sector {}",
+                update.deadline, update.sector_number
+            );
+            continue;
+        }
+
+        // This inefficiently loads deadline/partition info for each update.
+        if !state
+            .check_sector_active(
+                &store,
+                update.deadline,
+                update.partition,
+                update.sector_number,
+                true,
+            )
+            .with_context_code(ExitCode::USR_ILLEGAL_ARGUMENT, || {
+                format!("error checking state of sector {}", update.sector_number)
+            })?
+        {
+            info!("sector isn't active, skipping sector {}", update.sector_number);
+            continue;
+        }
+
+        let sector_info = if let Ok(value) = sectors.must_get(update.sector_number) {
+            value
+        } else {
+            info!("failed to get sector, skipping sector {}", update.sector_number);
+            continue;
+        };
+
+        if !sector_info.deal_ids.is_empty() {
+            info!("cannot update sector with deals, skipping sector {}", update.sector_number);
+            continue;
+        }
+
+        let expected_proof_type = sector_info
+            .seal_proof
+            .registered_update_proof()
+            .context_code(ExitCode::USR_ILLEGAL_STATE, "couldn't load update proof type")?;
+        if update.update_proof_type != expected_proof_type {
+            info!("unsupported update proof type {}", i64::from(update.update_proof_type));
+            continue;
```

Comment:
nit: For a future where this is called by a snark pack aggregate super snap method -- Im suspicious we might want to just fail the whole thing early here.  Don't think you need to do anything in this PR.